pkg/util/mount/nsenter_mount.go

```diff
@@ -281,49 +273,179 @@ func (mounter *NsenterMounter) MakeFile(pathname string) error {
+// links on the host. In other words, it does the same as filepath.EvalSymlinks()
+// would do when it ran on the host and not inside a container.
+//
+// mustExist makes evalHostSymlinks to return error when the path does not
+// exist. When it's false, it evaluates symlinks of the existing part and
+// blindly adds the non-existing part:
+// path: /mnt/volume/non/existing/directory. /mnt/volume exists and
+// non/existing/directory does not exist. It resolves symlinks in /mnt/volume
+// to say /mnt/foo and returns /mnt/foo/non/existing/directory.
+func (mounter *NsenterMounter) evalHostSymlinks(path string, mustExist bool) (string, error) {
+	// call /bin/realpath <opt> <path>.
+	realpath := mounter.ne.AbsHostPath("realpath")
+	var opt string
+	if mustExist {
+		// "realpath -e: all components of the path must exist"
+		opt = "-e"
+	} else {
+		// "realpath -m: no path components need exist or be a directory"
+		opt = "-m"
+	}
 
-	// There is no action when the container starts. Bind-mount will be cleaned
-	// when container stops by CleanSubPaths.
-	cleanupAction = nil
-	return newHostPath, cleanupAction, err
+	cmd := mounter.ne.Exec(realpath, []string{opt, path})
+	outBytes, err := cmd.CombinedOutput()
+	out := strings.TrimSpace(string(outBytes))
+	if err != nil {
+		return "", fmt.Errorf(out) // Typically "realpath: /mnt/blabla: No such file or directory"
+	}
+	return out, nil
 }
 
-func (mounter *NsenterMounter) SafeMakeDir(pathname string, base string, perm os.FileMode) error {
-	return doSafeMakeDir(pathname, base, perm)
+func doNsEnterBindSubPath(mounter *NsenterMounter, subpath Subpath) (hostPath string, err error) {
+	// Linux, kubelet runs in a container:
+	// - safely open the subpath
+	// - bind-mount the subpath to target (this can be unsafe)
+	// - check that we mounted the right thing by comparing device ID and inode
+	//   of the subpath (via safely opened fd) and the target (that's under our
+	//   control)
+
+	// Evaluate all symlinks here once for all subsequent functions.
+	newVolumePath, err := mounter.evalHostSymlinks(subpath.VolumePath, true /*mustExist*/)
+	if err != nil {
+		return "", fmt.Errorf("error resolving symlinks in %q: %v", subpath.VolumePath, err)
+	}
+	newPath, err := mounter.evalHostSymlinks(subpath.Path, true /*mustExist*/)
+	if err != nil {
+		return "", fmt.Errorf("error resolving symlinks in %q: %v", subpath.Path, err)
+	}
+	glog.V(5).Infof("doBindSubPath %q (%q) for volumepath %q", subpath.Path, newPath, subpath.VolumePath)
+	subpath.VolumePath = mounter.getKubeletPath(filepath.Clean(newVolumePath))
+	subpath.Path = mounter.getKubeletPath(filepath.Clean(newPath))
+
+	// Check the subpath is correct and open it
+	fd, err := safeOpenSubPath(mounter, subpath)
+	if err != nil {
+		return "", err
+	}
+	defer syscall.Close(fd)
+
+	glog.Infof("JSAF before prepare")
+	alreadyMounted, bindPathTarget, err := prepareSubpathTarget(mounter, subpath)
```

Comment:
should this be done first?

Reply:
IMO it does not really matter if we do `safeOpenSubPath` or `prepareSubpathTarget` first, I wanted the `defer` that cleans up volumes to affect as little code as possible.